docs/source/services.md

```diff
@@ -0,0 +1,84 @@
+# JupyterHub services
+
+JupyterHub 0.7 adds the notion of Services.
+A Service is a process that interacts with the Hub REST API.
+Services may perform actions such as shutting down user servers that have been idle for some time,
+or registering additional web servers that should also use the Hub's authentication
+and be served behind the Hub's proxy.
+
+There are two main characteristics of services:
+
+1. Is it **managed** by JupyterHub?
+2. Does it have a web server that should be added to the proxy?
+
+If a `command` is specified for launching the service, it will be started and managed by the Hub.
+If a `url` is specified for where the service runs its own webserver,
+it will be added to the Hub's proxy at `/service/:service-name`.
+
+## Managed services
+
+**Managed** services are services that the Hub starts and is responsible for.
+These can only be local subprocesses of the Hub.
+While there are similarities with notebook Spawners,
+there are no plans to support the same spawning abstractions as notebook
+If you want to run these services in docker or other environments,
```

Comment:
We've so far found it useful to have singleton spawning in the hub itself, since that lets it participate on the hub config system.  This makes it easy, for example, to run only sub-parts of our full hub setup (e.g. not using pgbouncer when running locally, or not creating and destroying heavyweight containers between development runs).

Currently what we're doing is overriding the hub's setup and teardown logic to first invoke our own setup code before continuing.  There are a couple things we're hooking this way, but the singleton container code looks like this:

```python
    @gen.coroutine
    def init_spawners(self):
        self.init_singleton_containers()
        yield super().init_spawners()

    def init_singleton_containers(self):
        """
        Initialize singleton containers and remove any single-user containers
        that are left over from the last run of the hub.
        """
        log = self.log
        log.info("Initializing singleton containers.")

        self.cleanup_single_user()

        # The client instantiated in each container doesn't survive getting
        # copied from the config file, so share our client among
        # the singleton containers.
        for container in self.singleton_containers:
            log.info("Starting [%s]" % container.name)
            instances = container.instances()
            # Remove any existing instances of this container before starting.
            if instances and self.clean_singletons:
                log.warn("Found instance of [%s]. Purging." % container.name)
                container.purge()
            elif not instances:
                container.run()

    def cleanup_containers(self):
        """
        Clean up singleton containers and any single-user containers that
        haven't been removed yet.
        """
        self.log.info("Cleaning up singleton containers.")
        if self.research_config.clean_singletons:
            for container in self.singleton_containers:
                self.log.info("Stopping singleton: [%s]" % container.name)
                container.purge(stop_first=True, remove_volumes=True)

        self.cleanup_single_user()
```

This has worked well enough for us so far, so I'm not super concerned that services won't support our use -case.  I do, however, think there's a useful class of services which are conceptually like local processes but require some other startup mechanism.  Another example of this might be the ability to set up new `PeriodicCallback`s to start and stop with the hub (we do this to make our hubs send heartbeats to an external monitoring system).